### PR TITLE
Angelina

### PR DESCRIPTION
**Pull request by Angelina-Yaroshenko**
added loss and train for it.

### CODE REVIEW

File: experiments/train_glass.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+import numpy as np
+import tqdm
+
+import shutil
+import argparse
+import os
+from typing import Tuple
+
+class net_2last_layers(torch.nn.Module):
```
Comment (svdcvt) [inline experiments/train_glass.py]: классы называть надо с большой буквы + как-то странно через нижнее подчеркивание

File: experiments/train_glass.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    def __init__(self,mod):
```
Comment (svdcvt) [inline experiments/train_glass.py]: я не хочу щас везде править пробелы и прочее по PEP8, но я уже вижу много

File: experiments/train_glass.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        labels = torch.cat(all_labels)
+        log_dict_with_writer(labels, logits, train_writer, ood_label=ood_label)
+        print("  Evaluation results: \n   Accuracy: {:.4f}\n   Loss: {:.4f}".format(accuracy, loss))
+    return loss, accuracy
+
+
+def train(**kwargs):
+    
+    
+    parser = argparse.ArgumentParser()
```
Comment (svdcvt) [inline experiments/train_glass.py:92]: посмотри в ветке руслана мы отказались от аргпарса и делаем конфиги, наверное будет проще тебе сразу это поправить и уже это будет все норм в мастере

File: experiments/train_glass.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    
+    
+    train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True, drop_last=True)
+    val_dataloader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False, drop_last=False)
+    # if torch.cuda.is_available(): TODO
+    #    train_dataloader = DataPrefetcher(train_dataloader)
+    #    val_dataloader = DataPrefetcher(val_dataloader)
+    #    pass
+#    criterion = nn.CrossEntropyLoss()
+    criterion = glass_loss(alpha = 10.)
```
Comment (svdcvt) [inline experiments/train_glass.py]: а альфа это гиперпараметр? если да то надо его в конфиг засунуть

File: src/food/ood/loss.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import torch
+import torch.nn.functional as F
+
+
+class glass_reg(torch.nn.Module):
```
Comment (svdcvt) [inline src/food/ood/loss.py]: тот же самый момент с названием, назови например GlassReg

File: src/food/ood/loss.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    def forward(self, target: torch.Tensor, prediction: torch.Tensor) -> torch.Tensor:
+        L = torch.Tensor([0]).to(target.device)
+        c = 0.1
+        for i in range(target.shape[0]):
+            l = torch.max((prediction[i] - prediction[i][target[i]] + c).T,torch.zeros_like(prediction[i]))
+            l[target[i]] = 0
+            L+= l.sum()  # added pred[ind]-pred[ind]+c 
+        return L/prediction.shape[0]
+        
+class glass_loss(torch.nn.Module):
```
Comment (svdcvt) [inline src/food/ood/loss.py]: GlassLoss

File: src/food/ood/loss.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        ind = diag.nonzero(as_tuple = True)
+        diag[ind] = 1./ diag[ind]
+        Z = torch.diag(diag)
+
+        reg = torch.matmul(V.weight,V.weight.T) - (torch.matmul(A,Z)+torch.matmul(Z,A))/2
+        L=torch.norm(reg,p='fro')/n_classes**2
+
+        return L
+    
+class glass_simple_loss(torch.nn.Module):
```
Comment (svdcvt) [inline src/food/ood/loss.py]: GlassLossSimple

File: src/food/ood/loss.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    def forward(self, model: torch.nn.Module, V: torch.Tensor, target: torch.Tensor,
+            prediction: torch.Tensor, logits: torch.Tensor) -> torch.Tensor:
+        mask = target<V.weight.shape[0]
+        self._loss = self._loss.to(target.device)
+        loss = self._loss(logits[mask,:],target[mask])
+        reg = self._reg(model, V, target[mask])
+        
+        return loss +self._alpha * reg  
+
```
Comment (svdcvt) [inline src/food/ood/loss.py]: общий комментарий 
1) будет круто если будет поясняющая документация хотя бы кратко к новым штукам, что за класс, что делает
2) проверить техническую часть пока не могу, потому что не шарю
3) правь пеп

**Review by svdcvt — CHANGES_REQUESTED** [review]

File: src/food/ood/loss.py (hunk 5 above)
Comment (new-okaerinasai) [reply src/food/ood/loss.py]: Назови файл не просто "loss", а более информативно -- у тебя же там не просто лосс, а именно GLaS.

File: src/food/ood/loss.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        target_sparse = target_sparse.index_select(0, target)
+
+        A = torch.matmul(target_sparse.T,target_sparse)
+        diag = torch.diag(A)
+        ind = diag.nonzero(as_tuple = True)
+        diag[ind] = 1./ diag[ind]
+        Z = torch.diag(diag)
+
+        reg = torch.matmul(V.weight,V.weight.T) - (torch.matmul(A,Z)+torch.matmul(Z,A))/2
+        L=torch.norm(reg,p='fro')/n_classes**2
```
Comment (new-okaerinasai) [inline src/food/ood/loss.py]: Поправь ВЕЗДЕ по pep8. Код не особо читаемый. Можно воспользоваться autopep8. Я внесу этот момент в CI.
```suggestion
        L = torch.norm(reg,p='fro' ) / n_classes ** 2
```

File: experiments/train_glass.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
+import food
```
Comment (new-okaerinasai) [inline experiments/train_glass.py:1]: Общее замечание: к своему коду обязательно напиши тесты. Когда мы ревьювим, можем проглядеть какие-то моменты. Как их правильно писать, расскажу.

**Review by svdcvt — APPROVED** [review]

**Review by new-okaerinasai — APPROVED** [review]
